cmd/soroban-rpc/internal/jsonrpc.go

```diff
@@ -223,13 +251,32 @@ func NewJSONRPCHandler(cfg *config.Config, params HandlerParams) Handler {
 		Namespace: params.Daemon.MetricsNamespace(), Subsystem: "network", Name: "global_inflight_requests",
 		Help: "Number of concurrenty in-flight http requests",
 	})
+
+	queueLimitedBridge := network.MakeHTTPBacklogQueueLimiter(
+		bridge,
+		globalQueueRequestBacklogLimiter,
+		uint64(cfg.RequestBacklogGlobalQueueLimit),
+		params.Logger)
+
+	globalQueueRequestExecutionDurationWarningCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: params.Daemon.MetricsNamespace(), Subsystem: "network", Name: "global_request_execution_duration_threshold_warning",
+		Help: "The metric measures the count of requests that surpassed the warning threshold for execution time",
+	})
+	globalQueueRequestExecutionDurationLimitCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: params.Daemon.MetricsNamespace(), Subsystem: "network", Name: "global_request_execution_duration_threshold_limit",
+		Help: "The metric measures the count of requests that surpassed the limit threshold for execution time",
+	})
+	durationLimitedBridge := network.MakeHTTPRequestDurationLimiter(
+		queueLimitedBridge,
+		cfg.RequestExecutionWarningThreshold,
+		cfg.MaxRequestExecutionDuration,
+		globalQueueRequestExecutionDurationWarningCounter,
+		globalQueueRequestExecutionDurationLimitCounter,
+		params.Logger)
+
 	return Handler{
-		bridge: bridge,
-		logger: params.Logger,
-		Handler: network.MakeHTTPBacklogQueueLimiter(
-			bridge,
-			globalQueueRequestBacklogLimiter,
-			uint64(cfg.RequestBacklogGlobalQueueLimit),
-			params.Logger),
+		bridge:  bridge,
+		logger:  params.Logger,
+		Handler: durationLimitedBridge,
```

Comment:
is there enough value in having the extra `httpRequestDurationLimiter` to warrant it's additional implementation and complexity, it's mainly effective as a blanket override or stop-gap on top of the per-jrpc endpoint limits/thresholds/warns implemented by `rpcRequestDurationLimiter`? from naive outside observation, it just appears to add a significant amount of code including some duplication for marginal gain, but I might be missing the nuance of it.

Reply:
Great question ! I was thinking about that topic as well.

The major reason that this is beneficial is due to a jrpc feature called "batching", which allow the caller to "pack" multiple calls within a single http call.

Limiting the total execution time per session gives us better opportunity to evenly distribute the available server resources across the users.

Note that http calls *can* be easily throttled and load balanced. This means that by "pushing" users to avoid using big packed jrpc batches, we're allowing ourselves to provide better scaling by taking advantage of http-aware load balancer.